compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/handlers/StepHandler.kt

```diff
@@ -0,0 +1,290 @@
+import org.jetbrains.kotlin.ir.types.isInt
+import org.jetbrains.kotlin.ir.types.isLong
+import org.jetbrains.kotlin.ir.types.toKotlinType
+import org.jetbrains.kotlin.ir.types.typeWith
+import org.jetbrains.kotlin.ir.visitors.IrElementVisitor
+import org.jetbrains.kotlin.name.FqName
+import kotlin.math.absoluteValue
+
+/** Builds a [ProgressionHeaderInfo] for progressions built using the `step` extension function. */
+internal class StepHandler(
+    private val context: CommonBackendContext,
+    private val visitor: IrElementVisitor<HeaderInfo?, Nothing?>
+) : ProgressionHandler {
+
+    private val symbols = context.ir.symbols
+
+    override val matcher = SimpleCalleeMatcher {
+        singleArgumentExtension(FqName("kotlin.ranges.step"), symbols.progressionClasses.map { it.typeWith() })
+        parameter(0) { it.type.isInt() || it.type.isLong() }
+    }
+
+    override fun build(expression: IrCall, data: ProgressionType, scopeOwner: IrSymbol): HeaderInfo? =
+        with(context.createIrBuilder(scopeOwner, expression.startOffset, expression.endOffset)) {
+            // Retrieve the HeaderInfo from the underlying progression (if any).
+            val nestedInfo = expression.extensionReceiver!!.accept(visitor, null) as? ProgressionHeaderInfo
+                ?: return null
+
+            val stepArg = expression.getValueArgument(0)!!
+            // We can return the nested info if its step is constant and its absolute value is the same as the step argument. Examples:
+            //
+            //   1..10 step 1               // Nested step is 1, argument is 1. Equivalent to `1..10`.
+            //   10 downTo 1 step 1         // Nested step is -1, argument is 1. Equivalent to `10 downTo 1`.
+            //   10 downTo 1 step 2 step 2  // Nested step is -2, argument is 2. Equivalent to `10 downTo 1 step 2`.
+            if (stepArg.constLongValue != null && nestedInfo.step.constLongValue?.absoluteValue == stepArg.constLongValue) {
+                return nestedInfo
+            }
+
+            // To reduce local variable usage, we create and use temporary variables only if necessary.
+            val (stepArgVar, stepArgExpression) = createTemporaryVariableIfNecessary(stepArg, "stepArg")
+
+            // The `step` standard library function only accepts positive values, and performs the following check:
+            //
+            //   if (step > 0) step else throw IllegalArgumentException("Step must be positive, was: $step.")
+            //
+            // We insert this check in the lowered form only if necessary.
+            val stepType = data.stepType(context.irBuiltIns)
+            val stepGreaterFun = context.irBuiltIns.greaterFunByOperandType[stepType.toKotlinType()]!!
+            val zeroStep = if (data == ProgressionType.LONG_PROGRESSION) irLong(0) else irInt(0)
+            val throwIllegalStepExceptionCall = irCall(context.irBuiltIns.illegalArgumentExceptionSymbol, stepType).apply {
```

Comment:
Maybe move the construction of this to the case where it is needed to not build the IR for it when it is not needed?

Reply:
Easy enough, converted to lambda